Refactor COM Registration Fixing Tool part 2: remove bugs, return Windows errors, improve UX

XLTechie opened the pull request:
### Link to issue number:

Fixes #10799
Fixes #12345 
Fixes #12351 
Replaces #12349

### Summary of the issue:

* The COM Registration Fixing Tool still showed its completion dialog if it failed, or if the user said no to the UAC dialog (#12345).
* The initial dialog didn't have a visual close button (red X), and was not responsive to the `Escape` or `Alt+F4` keys (#10799).
* It didn't report errors to the user as pointed out by @CyrilleB79 in #12349, and
* Its initial warning dialog was not serving users well with its strong warning and presentation (#12351).

### User facing changes:

* Initial dialog given a more friendly message, with a more beginner-themed explanation.
* User may now press `Escape`/`Alt+F4` to leave the dialog, as well as hitting `Cancel`.
* There is now a visual close control (red X).
* `F1` context help is now available.
* Errors are reported to the user by error code.
* The tool no longer reports success if it fails.
* The tool no longer reports success if the process is cancelled by UAC.

### Development details:

* Refactored this code to be in line with current coding standards, fix the above issues, and to give the tool a better over all UX.
* Catch the Windows error generated on UAC cancel, and log that the process was stopped at the UAC.
* Show other Windows errors to the user via a dialog, instead of just in the log.
* Switched from using YES and NO buttons, to Continue and CANCEL.
* Added a docstring. Linted and added more logging.
* In order to implement the Continue and Cancel paradigm, created `gui.nvdaControls._ContinueCancelDialog`.

### Testing strategy:

* Tested all buttons by tabbing and pressing, and by keyboard shortcuts.
* Tested that all windows appear when and where they should.
* Confirmed that the "completed" message no longer appears on failure or cancellation.
* Simulated a Windows error to prove that a message would appear to show the error to the user.

### Known issues with pull request:

1. Upon UAC cancel, or upon successful completion, focus is moved from desktop or where ever it was, to the NVDA systray icon. That seems a little strange to me, although it is not new behavior. `gui.postPopup` is supposed to fix this, but it doesn't.
1. Even after being deleted, the progress window's speech appears to hang around until after the completion window is cleared, although this speech seems to be cancelled or interrupted. Again, not new behavior. Log fragment showing this below:

```
IO - speech.speak (02:07:07.616) - MainThread (8368):
Speaking ['COM Registration Fixing Tool', 'dialog', 'The COM Registration Fixing Tool has completed successfully.', CancellableSpeech (still valid)]
IO - speech.speak (02:07:07.621) - MainThread (8368):
Speaking ['OK', 'button', CancellableSpeech (still valid)]
IO - inputCore.InputManager.executeGesture (02:07:09.326) - winInputHook (6992):
Input: kb(desktop):enter
IO - speech.speak (02:07:09.406) - MainThread (8368):
Speaking ['COM Registration Fixing Tool', 'dialog', "Please wait while NVDA attempts to fix your system's COM registrations...", CancellableSpeech (still valid)]
IO - speech.speak (02:07:09.411) - MainThread (8368):
Speaking ['COM Registration Fixing Tool', CancellableSpeech (still valid)]
```

### Code Review Checklist:

- [X] Pull Request description is up to date.
- [X] Unit tests.
- [X] System (end to end) tests.
- [X] Manual tests.
- [X] User Documentation.
- [X] Change log entry.
- [ ] Context sensitive help for GUI changes.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced a custom dialog with "Continue" and "Cancel" buttons for better user interaction.

- **User Experience Improvements**
	- Updated messaging in the COM Registration Fixing Tool for better clarity and guidance.
	- Enhanced error handling and logging for a more reliable user experience.
	- Users can now exit the initial window using the `escape` key or `alt+f4`.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->
<!-- Edited by @XLTechie, since @coderabbitai had a bunch of unrelated junk in the summary. -->

michaelDCurran commented:
What is the status of this pr now that pr #12560  has been merged? Has this one been replaced, or is there extra work in this pr?

leonardder commented:
@XLTechie any chance you can continue this any time soon?

XLTechie commented:
I'll try to evaluate it next month.

CyrilleB79 commented:
Ping @XLTechie:
Do you plan to push forward this PR soon?

More specifically, I was about to implement a fix for #12345 when I remembered of your PR.

XLTechie commented:
Planning to, yes. Tried a couple weeks ago, but ran into a problem with my 
compile setup.

CyrilleB79 commented:
For issues with compile setup, feel free to ask on nvda-devel mailing list where such issues are usually discussed and addressed.

seanbudd commented:
@XLTechie - do you plan to finish this work? Otherwise I think it should be closed

XLTechie commented:
@seanbudd yes. I put it aside because I kept running into wx issues with text formatting being very strange. I'll look at it again, and if I can't make it work as-is, I'll split it into separate PRs addressing the major portions, so the important bits can get merged.

Sorry for letting it lay for so long--time flies.

seanbudd commented:
No problem - draft PRs are totally fine - we just want to make sure theyre not fully abandoned

Commit:
Refactored the COM Registration Fixing Tool to fix bugs and improve UX.

    gui.onRunCOMRegistrationFixesCommand:
    - No longer shows success dialog on failure, or when the UAC is canceled or declined (#12345)..
    - We now catch the Windows error that signals user cancel of UAC, and return as if "NO" had been initially chosen.
    - If there is a Windows error other than cancel of UAC, alert the user and show the error in a dialog.
    - Rewrote initial dialog message to be more friendly and helpful, and removed warning icon (#12351).
    - Switched from using YES and NO buttons, to OK and Cancel (CANCEL button provides visual closure and escape key use).
    - Added a docstring and fully linted code. Added more debug logging.

Commit:
changes

pre-commit-ci[bot] committed:
Pre-commit auto-fix

Commit:
String reformatting

Commit:
Remove old imports

Commit:
Emitintro message using a method.

Commit:
Fix quoting

pre-commit-ci[bot] committed:
Pre-commit auto-fix

Commit:
Line length fixes.

XLTechie commented on source/gui/__init__.py:
```diff
@@ -483,49 +483,90 @@ def onInstallCommand(self, evt):
+		"""
+		# Translators: The title of various dialogs displayed when using the COM Registration Fixing tool
+		genericTitle: str = _("Fix COM Registrations")
+		response: int = messageBox(
```
@coderabbitai Can you suggest code for replacing `messageBox` in this usage, with a custom dialog based on `gui.nvdaControls.MessageDialog`? When doing so, override `MessageDialog`'s `_addButtons`, to provide `Continue` and `Cancel` buttons instead of `OK` and `Cancel` buttons.

XLTechie replied:
@coderabbitai this fails, because in `gui.nvdaControls.MessageDialog`, the implementation of `_addButtons` uses a `buttonHelper` argument, which you failed to include when overriding the method. Therefore the derivative class is incompatible with the superclass.

Commit:
Fix bad perms.

Commit:
Trying new continue/cancel dialog

Commit:
Fix various things the AI did wrong.

Commit:
Remove warning icon and ding.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

XLTechie commented:
@seanbudd the dev codingStandards file says, and it has been NVDA standard for a long time:

```
For the parameter list of function definitions, double indent, this differentiates the
    parameters and the body of the function.
```

I did so, with the new `utilityDialogs` module that I created here. And then @pre-commit-ci immediately un-did it.

Is codingStandards out of date?
Or is this the bot being tyrannical and forcing us to change our preferences to conform to its own?

Commit:
Fix title and message ordering in ContinueCancelDialog.__init__

Commit:
Remove unnecessary UTF-8 header line

XLTechie commented:
@coderabbitai Can you suggest a manual testing plan for the COM Registration Fixing Tool? Especially important is a method to test for error conditions, and whether error codes are shown to the user on failure. Suggest a way to simulate or cause a failure.

XLTechie commented:
@britechguy You may want to do a visual and editorial check of the initial dialog on this, since much of the text is from your suggested revision. A downloadable installer is available [here](https://appveyor-artifacts-enam.7b474ce6bd9813bd1e711f7cdc19151f.r2.cloudflarestorage.com/xltechie-128921/xlnvda/crftrefactor-357/3u8fu8se8m7nl5on/output/nvda_snapshot_CRFTRefactor-357%2C60d6a4fc.exe?X-Amz-Expires=300&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=23e97e8f2c472db9a7966a0f1cd5e0c6/20240903/us-east-1/s3/aws4_request&X-Amz-Date=20240903T105124Z&X-Amz-SignedHeaders=host&X-Amz-Signature=b006215c0ebfe244371645cfe86e536858751dd6b0adab4edbf7f8981710eb7b), or under the AppVeyor details above once the NV Access version finishes building.

Commit:
Fix instruction from "Press OK" to "Press Continue".

Commit:
Add docstring.

XLTechie commented:
@coderabbitai review

XLTechie commented on source/gui/utilityDialogs.py:
```diff
@@ -0,0 +1,47 @@
+		title: str,
+		message: str,
+		dialogType: int = MessageDialog.DIALOG_TYPE_STANDARD,
+		continueByDefault: bool = True,
```
I am aware of these points, but can not pinpoint the exact problem. Please do dive further into the codebase. The code which should be setting the `Cancel` button to default when this argument is False, is in the overridden method `_addButtons`.

XLTechie replied:
@coderabbitai the code you suggested is already there, in the `else` clause at the end of `_addButtons`.

XLTechie replied:
@seanbudd any idea why `SetDefault` is not working for cancel buttons here? I even tried hard-coding it to set the cancel button as default right after the button was created, and Continue is still set as default.

It's not actually relevant for this PR, but I was trying to make the dialog generic. I'll remove that feature if we can't figure it out. I appear to have broken the rabbit so no help there, and web searches on `SetDefault` failures haven't turned up anything useful yet.

Commit:
Fix-up some indenting.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

Commit:
Merge branch 'master' into CRFTRefactor

Commit:
Added a note to changes for devs.

britechguy commented:
Luke,

I would be happy to check whatever someone wants me to check.  That being
said, I cannot make heads nor tails of what it is I should be looking for
based on #12355 and all of the bits it leads back to.  I can't even figure
out where I became involved somewhere in the mix somewhere in the past.
This is one of the most "tangled web" issues I've ever encountered and I
can't tease out where anything I may have said at some point in the past is
or fits into the picture.

Brian

> Message ID: ***@***.***>
>

XLTechie commented:
@britechguy First, here is the [latest installer](https://ci.appveyor.com/api/buildjobs/o0536bdm681o5i11/artifacts/output%2Fnvda_snapshot_pr12355-33795%2C5ad99f71.exe).
The specific things I wanted to draw your attention toward, are:

1. The language of the first dialog shown when you run the COM Registration Fixing Tool, pursuant to [this comment](https://github.com/nvaccess/nvda/issues/12351#issuecomment-829344413) by you, and especially [this comment by you](https://github.com/nvaccess/nvda/issues/12351#issuecomment-829529714), on which I primarily based the language. I don't know if it will survive review, but I credited you in the changelog for that one.
2. The visual closure (red X) of that same first dialog should now be possible. You weren't involved with that issue, but are well positioned to comment on the UX.

XLTechie commented:
@coderabbitai If the initial dialog of the COM Registration Fixing Tool is cancelled, NVDA focus returns to whatever application had focus before the NVDA menu was opened to start the tool. However if the tool run is allowed to complete, after the completion message is shown, focus is sent to the NVDA icon in the notification area. How can we cause focus to return to the initial application in the second scenario?

XLTechie commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
+						# Translators: message shown to the user on COM Registration Fix fail
+						"The COM Registration Fixing Tool was unsuccessful. This Windows "
+						"error may provide more information.    {}",
+					).format(error),
+					# Translators: Added to the title of the dialog showing the COM Registration Fix failure
+					genericTitle + " " + _("Failed"),
+					wx.OK,
+				)
+		# Display success dialog if there were no errors
 		messageBox(
 			_(
-				# Translators: The message displayed when the COM Registration Fixing tool completes.
-				"The COM Registration Fixing tool has finished. "
+				# Translators: Message shown when the COM Registration Fixing tool completes.
+				"The COM Registration Fixing Tool has completed successfully.\n"
 				"It is highly recommended that you restart your computer now, to make sure the changes take full effect.",
 			),
-			# Translators: The title of a dialog presented when the COM Registration Fixing tool is complete.
-			_("COM Registration Fixing Tool"),
+			genericTitle,
 			wx.OK,
 		)
 
+	def _CRFTIntroMessage(self) -> str:
+		"""Helper that returns the COM Registration Fixing Tool's introductory statement."""
+		return _(
+			# Translators: Explain the COM Registration Fixing tool to users before running
+			"Welcome to the COM Registration Fixing tool.\n"
+			"This tool is used by NVDA to fix problems it may have as it tries to interact with various applications, "
+			"or with Windows itself.\n"
+			"It examines the system registry for corrupted or missing accessibility entries and will correct them.\n"
+			"Those entries are sometimes damaged by installing or uninstalling programs, or other system events. This "
+			'can result in "unknown" or "pane" being spoken instead of the content you were expecting, or previously '
+			"accessible elements suddenly no longer reading correctly.\n\n"
+			"You have most likely been asked to run this tool by NVDA support or a power user trying to assist you.\n"
+			"Because it needs to modify the Windows registry, if you have User Account Control (UAC) active, you will "
+			"be prompted by UAC before this tool can do its job. This is normal and you should answer by pressing the "
+			"yes button.\n\n"
+			"Press Continue to try to repair the registry now.\n",
+		)
```
I believe this would only save four lines in the original method, and serves no real purpose other than symmetry with the intro message. If reviewers want it, I'll do it, but otherwise I don't think it really improves things.

Commit:
Merge branch 'master' into CRFTRefactor

XLTechie commented:
@coderabbitai resolve

Commit:
Implemented an alternate way of having the Cancel button be selected by default in gui.utilityDialogs.ContinueCancelDialog, if a caller prefers that.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

XLTechie commented on source/gui/utilityDialogs.py:
I put this in a new `gui.utilityDialogs` module. Not sure if that is useful. I could have put it in `nvdaControls`, but wasn't clear on the scope of what was meant to be in there. Happy to move it if desired.

SaschaCowley replied:
Could you move it to `nvdaControls` instead?

XLTechie commented on source/gui/utilityDialogs.py:
```diff
@@ -0,0 +1,58 @@
+		:param title: The title or caption of the dialog.
+		:param message: The message to be shown in the dialog.
+		:param dialogType: One of the dialog type constants from MessageDialog, defaults to standard.
+		:param continueButtonFirst: If True, the Continue button will appear first, and be selected when the dialog
```
I did it this way because I couldn't find another way. Originally I tried fixed positions for the two buttons, and just using `SetDefault()` on the `Cancel` button, if the caller asked for that. However I could never get `SetDefault` to be honored in the overridden `_addButtons` method, and could not figure out why it was being ignored.

I found, however, that whichever button was added first, became the default. Since I wanted this dialog to be somewhat generic, I switched to doing this. I put a comment to that effect in `_addButtons`.

XLTechie commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
+					wx.OK,
+				)
+		# Display success dialog if there were no errors
 		messageBox(
```
The AI wanted me to also move this out into a helper method, as it asked me to do with the intro message. It made sense for the very long intro message, but IMO it is pointless for any of these other short messages.
Happy to do it if reviewers prefer though.

XLTechie reviewed (commented):
@seanbudd this is ready for review I believe. I dropped a few comments to reviewers on specific lines below for questionable items.

XLTechie commented:
@coderabbitai Can you check the updated code for any missing or inaccurate type annotations on variables or methods, and make GitHub committable suggestions to fix any defects?

XLTechie commented on source/gui/utilityDialogs.py:
```diff
@@ -0,0 +1,58 @@
+	These serve the same functions as `OK` and `Cancel` in other dialogs, but may be more desirable in some situations.
+	"""
+
+	def __init__(
```
@seanbudd I had these double indented as per the codingStandards file, but Pre-commit immediately reduced them to single indenting. Not sure what to do about that, if anything.

seanbudd replied:
I think we need to update codingStandards.md

britechguy commented:
Luke,

I have finally managed to get to the tests you requested.

I am perfectly fine with the reworded dialog box for the COM Registration
Fixing Tool.  It is far less alarming to the uninitiated than the original
was.

I am able to close both the first COM Registration Fixing Tool and the
second completion dialog with a point and click of "the red X" in its
typical location.  I have also tested using Escape as well as Alt + F4 on
the first dialog, both of which close it as I would expect for almost any
Windows dialog.

Brian

On Tue, Sep 3, 2024 at 9:59 PM Luke Davis ***@***.***> wrote:

> @britechguy <https://github.com/britechguy> First, here is the latest
> installer
> <https://ci.appveyor.com/api/buildjobs/o0536bdm681o5i11/artifacts/output%2Fnvda_snapshot_pr12355-33795%2C5ad99f71.exe>
> .
> The specific things I wanted to draw your attention toward, are:
>
>    1. The language of the first dialog shown when you run the COM
>    Registration Fixing Tool, pursuant to this comment
>    <https://github.com/nvaccess/nvda/issues/12351#issuecomment-829344413>
>    by you, and especially this comment by you
>    <https://github.com/nvaccess/nvda/issues/12351#issuecomment-829529714>,
>    on which I primarily based the language. I don't know if it will survive
>    review, but I credited you in the changelog for that one.
>    2. The visual closure (red X) of that same first dialog should now be
>    possible. You weren't involved with that issue, but are well positioned to
>    comment on the UX.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/12355#issuecomment-2327759920>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AENFMHSXEAJTBLMDZPDOX4LZUZSO3AVCNFSM433Y7ZOKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMZSG43TKOJZGIYA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

XLTechie commented:
Thanks for testing, Brian.

seanbudd commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
+			"Those entries are sometimes damaged by installing or uninstalling programs, or other system events. This "
+			'can result in "unknown" or "pane" being spoken instead of the content you were expecting, or previously '
+			"accessible elements suddenly no longer reading correctly.\n\n"
+			"You have most likely been asked to run this tool by NVDA support or a power user trying to assist you.\n"
```
I feel like "NVDA support" is not very well defined and "power user" might be an unfamiliar term to novice users.
I don't have a good suggestion, but perhaps something like

```suggestion
			"You have most likely been asked to run this tool by an experienced user or support worker to investigate and potentially fix an issue with NVDA's behaviour.\n"
```

XLTechie committed:
Apply Sean's proposed language to replace the "...power user..." reference.

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

Commit:
Merge branch 'master' into CRFTRefactor

Commit:
One more try at NVDA style method definition indenting.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

XLTechie commented:
@seanbudd I applied your language change suggestion as-is, it looked like a reasonable rewording to me. I'll open a PR to change codingStandards to fit what the Pre-commit bot wants.

SaschaCowley commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
+						"error may provide more information.    {}",
+					).format(error),
+					# Translators: Added to the title of the dialog showing the COM Registration Fix failure
+					genericTitle + " " + _("Failed"),
```
I think we should create a different title string altogether, as we don't know that `<action>+<failed>` is the best way of titling the dialog in all languages.

XLTechie replied:
```suggestion
					_("COM Registration Fixing Tool Failed"),
```

SaschaCowley commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
 			wx.OK,
 		)
 
+	def _CRFTIntroMessage(self) -> str:
+		"""Helper that returns the COM Registration Fixing Tool's introductory statement."""
+		return _(
+			# Translators: Explain the COM Registration Fixing tool to users before running
+			"Welcome to the COM Registration Fixing tool.\n"
+			"This tool is used by NVDA to fix problems it may have as it tries to interact with various applications, "
+			"or with Windows itself.\n"
+			"It examines the system registry for corrupted or missing accessibility entries and will correct them.\n"
+			"Those entries are sometimes damaged by installing or uninstalling programs, or other system events. This "
+			'can result in "unknown" or "pane" being spoken instead of the content you were expecting, or previously '
+			"accessible elements suddenly no longer reading correctly.\n\n"
+			"You have most likely been asked to run this tool by an experienced user or support worker to investigate and potentially fix an issue with NVDA's behaviour.\n"
+			"Because it needs to modify the Windows registry, if you have User Account Control (UAC) active, you will "
+			"be prompted by UAC before this tool can do its job. This is normal and you should answer by pressing the "
+			"yes button.\n\n"
+			"Press Continue to try to repair the registry now.\n",
+		)
```
Can this be made an attribute rather than a method?

SaschaCowley commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
+			"accessible elements suddenly no longer reading correctly.\n\n"
+			"You have most likely been asked to run this tool by an experienced user or support worker to investigate and potentially fix an issue with NVDA's behaviour.\n"
+			"Because it needs to modify the Windows registry, if you have User Account Control (UAC) active, you will "
+			"be prompted by UAC before this tool can do its job. This is normal and you should answer by pressing the "
+			"yes button.\n\n"
```
Users whose account doesn't have administrative privileges will need to enter the admin password. These instructions should either be modified to account for this, or removed entirely.

XLTechie replied:
```suggestion
			"be prompted by UAC before this tool can do its job. This is normal and you should follow the UAC "
			"instructions to approve the running of the tool.\n\n"
```

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -18,13 +18,18 @@ In order to use this feature, the application volume adjuster needs to be enable
 * The Report link destination, Character formatting information, and Speak selection dialogs, now include "Close" and "Copy" buttons for user convenience. (#17018, @XLTechie)
 * The exit dialog now allows you to restart NVDA with add-ons disabled and debug logging enabled simultaneously. (#11538, @CyrilleB79)r
+* Changes to the COM Registration Fixing Tool: (#12355, @XLTechie)
+  * It will now show a message to the user, including the error, in the rare event of a Windows error while attempting COM re-registrations.
+  * It now starts with a more user friendly explanation of its purpose, instead of a warning. (#12351, @XLTechie, @britechguy)
+  * The initial window can now be exited with `escape` or `alt+f4`. (#10799, @XLTechie)
```
I think these should be ordered in the order they are likely to be encountered.
```suggestion
  * It now starts with a more user friendly explanation of its purpose, instead of a warning. (#12351, @XLTechie, @britechguy)
  * The initial window can now be exited with `escape` or `alt+f4`. (#10799, @XLTechie)
  * It will now show a message to the user, including the error, in the rare event of a Windows error while attempting COM re-registrations.
```

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -45,6 +50,7 @@ Add-ons will need to be re-tested and have their manifest updated.
   * `BrowseModeTreeInterceptor` object has a new `getLinkTypeInDocument` method which accepts an URL to check the link type of the object
   * A `toggleBooleanValue` helper function has been added to `globalCommands`.
   It can be used in scripts to report the result when a boolean is toggled in `config.conf`
+* There is a new `gui.utilityDialogs.ContinueCancelDialog`, for producing simple dialogs with Continue and Cancel buttons. (#12355, @XLTechie)
```
Edit when moving `ContinueCancelDialog` to `gui.nvdaControls`.

XLTechie replied:
```suggestion
* There is a new `gui.nvdaControls.ContinueCancelDialog`, for producing simple dialogs with Continue and Cancel buttons. (#12355, @XLTechie)
```

SaschaCowley reviewed (commented):
These are great changes, @XLTechie! Mostly reorganisation requests from me.

XLTechie commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
+						"The COM Registration Fixing Tool was unsuccessful. This Windows "
+						"error may provide more information.    {}",
+					).format(error),
+					# Translators: Added to the title of the dialog showing the COM Registration Fix failure
```
```suggestion
					# Translators: The title of a COM Registration Fixing Tool dialog, when the tool has failed
```

XLTechie commented on source/gui/__init__.py:
```diff
@@ -483,49 +484,90 @@ def onInstallCommand(self, evt):
-			)
-			== wx.NO
-		):
+	def onRunCOMRegistrationFixesCommand(self, evt) -> None:
```
```suggestion
	def onRunCOMRegistrationFixesCommand(self, evt: wx.CommandEvent) -> None:
```

XLTechie committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

Commit:
Merge branch 'master' into CRFTRefactor

Commit:
Move new class to nvdaControls.py

Commit:
Transformed the intro message from a method to a constant.

XLTechie commented:
@SaschaCowley I believe I took care of all of your review points. I'll mark it for review again after the rebuild, when I get back this afternoon.

Commit:
Fix bad import.

SaschaCowley commented:
@XLTechie  is this ready for review?

XLTechie commented:
@SaschaCowley Yes.

seanbudd commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,94 @@ def onInstallCommand(self, evt):
+					_(
+						# Translators: message shown to the user on COM Registration Fix fail
+						"The COM Registration Fixing Tool was unsuccessful. This Windows "
+						"error may provide more information.    {}",
+					).format(error),
```
Better clarity for translators, better spacing rules

```suggestion
						"error may provide more information.\n{error}",
					).format(error=error),
```

XLTechie replied:
@seanbudd any idea why I can't apply this suggestion from the web? I've been noticing that a lot lately. Then other times the button is there like it should be. No logic that I've been able to find.

seanbudd commented on source/gui/nvdaControls.py:
```diff
@@ -2,6 +2,7 @@
 # Copyright (C) 2016-2024 NV Access Limited, Derek Riemer, Cyrille Bougot, Luke Davis, Leonard de Ruijter
```
could you explain the chmod for `source/gui/logViewer.py`?

XLTechie replied:
It was executable. Now it's not.

seanbudd commented on source/gui/nvdaControls.py:
```diff
@@ -376,6 +377,56 @@ def _onShowEvt(self, evt):
 		evt.Skip()
 
 
+class ContinueCancelDialog(MessageDialog):
```
Considering we want to amalgamate this into a single class via #13007 in this release, I think this should be a private class. I don't think we intend to release this dialog in 2025.1. @SaschaCowley - thoughts?

XLTechie replied:
I was unaware that was targeted for action soon. This can easily be made private.
I could even make it self-contained (from `MessageDialog`, at least), and change the implementation in some future PR.

seanbudd replied:
I appreciate it has stalled a lot - there's also a chance it gets stalled out of this release, but it is planned for 2025.1. I'll hold off on Sascha's thoughts on how to proceed here as she's picked up the work from me (very thankful for that)

SaschaCowley replied:
I support making this a private class. I think it worthwhile to keep it here so this work can be merged, just in case the message dialog API work doesn't end up making it to 2025.1. Also increases the test case surface for the new API, which is helpful.

seanbudd commented on source/gui/nvdaControls.py:
```diff
@@ -376,6 +377,56 @@ def _onShowEvt(self, evt):
+		self.continueButtonFirst: bool = continueButtonFirst
+		super().__init__(parent, title, message, dialogType)
+
+	def _addButtons(self, buttonHelper) -> None:
```
please add a type hint for buttonHelper for this function and the internal functions below

SaschaCowley commented on source/gui/__init__.py:
```diff
@@ -480,11 +480,32 @@ def onInstallCommand(self, evt):
+		"You have most likely been asked to run this tool by an experienced user or support worker to investigate "
+		"and potentially fix an issue with NVDA's behaviour.\n"
+		"Because it needs to modify the Windows registry, if you have User Account Control (UAC) active, you will "
+		"be prompted by UAC before this tool can do its job. This is normal and you should follow the UAC's "
+		"instructions to approve the running of the tool.\n\n"
```
Hmm, "follow the UAC's instructions" doesn't feel quite right to me for some reason. What do you think of
```suggestion
		"be prompted by UAC before this tool can do its job. This is normal, and you should follow the prompts to allow the tool to run.\n\n"
```

SaschaCowley commented on source/gui/__init__.py:
```diff
@@ -480,11 +480,32 @@ def onInstallCommand(self, evt):
+		"Because it needs to modify the Windows registry, if you have User Account Control (UAC) active, you will "
+		"be prompted by UAC before this tool can do its job. This is normal and you should follow the UAC's "
+		"instructions to approve the running of the tool.\n\n"
+		"Press Continue to try to repair the registry now.\n",
```
Why is there a line feed after the last line of this text?
```suggestion
		"Press Continue to try to repair the registry now.",
```

XLTechie replied:
@SaschaCowley I believe that is a remnant from when WX was causing the text to be mingled with the buttons for some reason in a previous implementation. I forgot that it wasn't needed with this one.

Commit:
Merge branch 'master' into CRFTRefactor

SaschaCowley commented on source/gui/__init__.py:
```diff
@@ -480,11 +480,32 @@ def onInstallCommand(self, evt):
 
 		installerGui.showInstallGui()
 
+	CRFT_INTRO_MSG: str = _(
```
Please expand "MSG" to "MESSAGE". I think we might as well make this private since I see no reason it needs to be public.
```suggestion
	CRFT_INTRO_MESSAGE: str = _(
```

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -52,7 +54,8 @@ Add-ons will need to be re-tested and have their manifest updated.
   * A `toggleBooleanValue` helper function has been added to `globalCommands`.
   It can be used in scripts to report the result when a boolean is toggled in `config.conf`
-* There is a new `gui.utilityDialogs.ContinueCancelDialog`, for producing simple dialogs with Continue and Cancel buttons. (#12355, @XLTechie)
+* There is a new `gui.nvdaControls.ContinueCancelDialog`, for producing simple dialogs with Continue and Cancel buttons. (#12355, @XLTechie)
```
Remove this when making the associated class private.

XLTechie commented on user_docs/en/changes.md:
```diff
@@ -49,6 +54,7 @@ Add-ons will need to be re-tested and have their manifest updated.
   * `BrowseModeTreeInterceptor` object has a new `getLinkTypeInDocument` method which accepts an URL to check the link type of the object
   * A `toggleBooleanValue` helper function has been added to `globalCommands`.
   It can be used in scripts to report the result when a boolean is toggled in `config.conf`
+* There is a new `gui.nvdaControls.ContinueCancelDialog`, for producing simple dialogs with Continue and Cancel buttons. (#12355, @XLTechie)
```
```suggestion
```

XLTechie committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>
Co-authored-by: Sean Budd <seanbudd123@gmail.com>

Commit:
Make ContinueCancelDialog private. Type hint ButtonHelper instances module-wide.

Commit:
Make message and class private in init.

Commit:
Fix missed change

Commit:
Merge branch 'master' into CRFTRefactor

Commit:
Rebuild

XLTechie commented:
@SaschaCowley @seanbudd This is ready again, I believe incorporating all of your changes.
The build hasn't started yet, but it builds fine on my AppVeyor setup.

Commit:
Edit user guide

SaschaCowley reviewed (approved):
This is looking pretty good to me. I think we should get @Qchristensen's input on the documentation changes.

XLTechie commented:
Thanks Sascha.

Qchristensen reviewed (approved):
Looks good, thanks Luke!

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -3799,19 +3799,25 @@ Follow the directions in [Creating a portable copy](#CreatingAPortableCopy) for
 
 Installing and uninstalling programs on a computer can, in certain cases, cause COM DLL files to get unregistered.
-As COM Interfaces such as IAccessible depend on correct COM DLL registrations, issues can appear in case the correct registration is missing.
+As accessibility Interfaces such as IAccessible depend on correct COM DLL registrations, issues can appear if the correct registrations are missing.
```
```suggestion
As accessibility interfaces such as IAccessible depend on correct COM DLL registrations, issues can appear if the correct registrations are missing.
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -3799,19 +3799,25 @@ Follow the directions in [Creating a portable copy](#CreatingAPortableCopy) for
+Because this tool corrects entries in the Windows registry, it requires administrative access to work.
+If you have UAC (User Access Control) enabled, as most users do, you will need to follow whatever prompts are presented by UAC, to run the tool successfully.
+This may include entering your administrative account's password.
+It will not be NVDA asking for this, but Windows itself, just as it would while installing a program.
```
I think this sentence is kind of misleading, NVDA is requesting UAC elevation

```suggestion
```

XLTechie replied:
@seanbudd The "asking for this" was referring to the password, not the whole UAC process, but I can see where confusion could enter in that wording.

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -3799,19 +3799,25 @@ Follow the directions in [Creating a portable copy](#CreatingAPortableCopy) for
+* NVDA is very slow when navigating in browsers while using browse mode.
 * And possibly other issues.
 
+Because this tool corrects entries in the Windows registry, it requires administrative access to work.
```
```suggestion
Because this tool corrects entries in the Windows registry, it requires administrative access to work, just like when installing a program.
```

CyrilleB79 commented:
I am commenting here quite late. Sorry of that.

Re-reading the text of this PR. I find it quite long and with lots of unnecessary information. Actually, I still prefer the current text than the proposal in this PR.

I acknowledge that some people may find the current dialog quite frightening. Though, minor modifications may make it acceptable:
* Replace the "Warning" title by "COM registration tool"
* Replace the Warning icon by a Question icon
* Start with "Welcome to the COM reg tool" as in this PR.

The text would become something like:
> Welcome to the COM Registration Fixing tool. This tool will try to fix common system problems that stop NVDA from being able to access content in various applications. This tool must make changes to the System registry and therefore requires administrative access.
> Press Continue to execute the tool or Cancel to exit it.

More specifically, when reading the new text, I have the following questions/ remarks:

> This tool is used by NVDA to fix problems it may have as it tries to interact with various applications, or with Windows itself.

* this tool is not "used by" NVDA, its provided by NVDA and used by the user. But this information is not useful.
* Why do we mention Windows itself? I guess that Windows interface elements such as taskbar or Windows explorer can be considered applications too? If not (maybe for the taskbar): "interact with various applications or with Windows interface" seems clearer.

> It examines the system registry for corrupted or missing accessibility entries and will correct them.
> Those entries are sometimes damaged by installing or uninstalling programs, or other system events. This can result in "unknown" or "pane" being spoken instead of the content you were expecting, or previously accessible elements suddenly no longer reading correctly.

IMO here, it'son only necessary to say that the tool will fix COM registration issues. Explaining how these problems may have appeared and what are the symptoms should be described only in the User Guide. It's not the goal of the GUI to explain all this. By the way, does context help work here?

> You have most likely been asked to run this tool by an experienced user or support worker to investigate and potentially fix an issue with NVDA's behaviour.

Why this? If there is no risk running this tool, this sentence is not useful. If this tool needs to be run only by experimented users or upon request in a support action, let's first describe clearly this risk (in this PR) and then decide which text should be written accordingly.

> Because it needs to modify the Windows registry, if you have User Account Control (UAC) active, you will be prompted by UAC before this tool can do its job. This is normal, and you should follow the prompts to allow the tool to run.

This part really seems the most exagerated. Everywhere else in NVDA, we only write "(requires administrative privileges)". We have not such description of the UAC mechanism for any other action requiring UAC, neither in NVDA's installer nor in the installer of any other application.

michaelDCurran commented:
While reading thoughts from  @CyrilleB79, I thought I should also state, just in case this was not clear to those working on / reviewing this pr, that the COM Registration Fixing Tool is always silently run within the NvDA installation process, every time NvDA is installed / updated.
 The reason why someone would run this tool separate to installation is if some other software was installed / uninstalled that broke the system, after NvDA was installed.

Commit:
Merge branch 'master' into CRFTRefactor

XLTechie commented:
@CyrilleB79 I think you make some valid points, and disagree with your comments on others. Rather than address each individually, I will make some updates to the text, and I ask you to re-read after the next commits.

One thing I will say, is that I think this overly friendly and descriptive text (mainly written by @britechguy, and with which I agreed), is in response to the unnecessary warning nature of the current dialog, which seems to want to make users think twice before running the tool.
I think the new text is more aimed at breaking the negative training that the current version of the dialog has given for several years.
I, as I'm sure Brian, and others, have seen users say they were afraid to run it, because they couldn't understand it, and the dialog essentially warned them not to, giving the impression that the tool is risky. That is frustrating, from a support prospective, when running the tool would have prevented their problem.

One thing I noticed in your quote, that I thought I had removed in an earlier commit, was the text that the CRFT "examines the system registry"--that isn't true, it doesn't examine anything, just brute-forces the state it wants. I will fix that again.

The other point I will make, is that you have asked before about context help for this dialog, and I delayed this PR a long time trying to find a way to do it.
That said, with the new dialog, there may be a way, I will try again.

Commit:
String update.

Commit:
Review suggestions for user guide

CyrilleB79 commented:
@XLTechie, I understand that the main scope of this PR is to make the CRFT more friendly so that people do not fear to run it anymore. No problem with that on my side.

I'll review the new wording and let you know if some questions remain.

Re context help. If it's not possible to support it immediately, let's not delay this PR for it. This PR should already ship with the rewording of the dialog and the bugfix.

Commit:
Attempt context help

pre-commit-ci[bot] committed:
Pre-commit auto-fix

XLTechie commented:
@CyrilleB79 Now that I have updated the text, I will comment on your statements that apply to the new version of the dialog:

>     * Replace the Warning icon by a Question icon

I didn't set any icon on the new dialog. I vaguely remember reading some Microsoft guidance that suggested this kind of dialog shouldn't.
I could be mistaken though.

>     * this tool is not "used by" NVDA, its provided by NVDA and used by the user. But this information is not useful.

As Mick pointed out, it is used by NVDA, at least during installation/update. But you're right that it is unnecessary to say so.

>     * Why do we mention Windows itself? I guess that Windows interface elements such as taskbar or Windows explorer can be considered applications too?

I doubt most users think of them so. We know how things like Explorer and various dialogs are implemented, but I imagine most users think of them as part of Windows.
I have changed that wording slightly.

> IMO here, it'son only necessary to say that the tool will fix COM registration issues.

Sorry, but 99% of users have no idea what a COM is, why it should be registered, or that it can have issues. Saying that would not help the people who most need it--i.e. someone who is having a problem, and searching the NVDA menus in frustration to see if there is some way to fix it.

> Explaining how these problems may have appeared and what are the symptoms should be described only in the User Guide.

I disagree. Most users aren't going to know what to look for in the user guide to find this further explained. If they go to the tool, and it gives some vague description of why someone might run it, they won't run it, thinking it probably doesn't apply to them. It's a repair tool. Not having it explain what it repairs, is just adding frustration to the user's already inaccessible and frustrating day.

But there is a bigger reason. Someone is most likely going to encounter this tool, when their already having a problem. They may have some vague idea that it's there, so they know to go find out when they have trouble, but remembering what it fixes, is a little too much to ask.
So consider that the user's web browser is just saying "pane" or "unknown". They remember the tool, so they go to run it, and read it's introduction, trying to figure out if they should run it.
Except we don't tell them why they might want to, and the only place they can find out--the user guide--will load in the web browser they can't use, which they came there to fix in the first place. Putting it only in the user guide is defeating the very people who need it most.

> This part *[About the UAC]* really seems the most exagerated. Everywhere else in NVDA, we only write "(requires administrative privileges)". We have not such description of the UAC mechanism for any other action requiring UAC, neither in NVDA's installer nor in the installer of any other application.

A few reasons, though I have simplified this text per your request.
This tool is really unprecedented in NVDA. Nothing else works like this does.
* The parts that just say "Requires administrative privileges", are controls in a settings dialog. Users know they are changing configurations by the nature of what they are doing, and dialogs full of settings aren't expected to contain long explanations.
* The reason it doesn't say anything when installing, is because users are expected to know that installing requires administrative privileges, as installing anything does. It comes with the nature of the action. If they say no, NVDA doesn't get installed, and the outcome is obvious. The outcome of not running this tool is not at all obvious. Why it should, or that it should, need administrative access, is also not at all obvious. It is very different than instillation. We can't make any assumptions about what the user knows when making this tool available, because there are very few other things like it, either in NVDA or Windows generally (apart from SFC and DISM, which are commandline Voodoo to most users who see them)

Commit:
Fix import

pre-commit-ci[bot] committed:
Pre-commit auto-fix

Adriani90 commented:
Actually I agree mostly with @CyrilleB79 in this regard. Adding lots of descriptive long texts to the GUI will overload it. I would rather provide a link to the user guide when opening the COM registration fixing tool dialog, advising users to read more information in the coresponding section.
Also an article about this in the in-process blog would have far more positive impact than a long descriptive text in the GUI, which probably doesn't fit on all kind of screens anyway.

Here is a suggestion for the introductory text for the dialog:

"Welcome to the COM Registration Fixing tool.
Running this tool will fix several problems with COM DLL files where NVDA reports "Unknown" or "pane"
because it cannot access the content of various applications and windows components.
Read more information about the COM registration fixing tool in the user guide, section 14.7.
Administrator priviledges are required to run the tool. Would you like to continue?"

zstanecic commented:
Priviledges > privileges

XLTechie commented:
@CyrilleB79 Thanks for prompting me to investigate context help again. With the new dialog type, it is now working.

Commit:
Merge branch 'master' into CRFTRefactor

Commit:
Docstring update for context help

Commit:
One more update of the message.

CyrilleB79 commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,90 @@ def onInstallCommand(self, evt):
+		"Welcome to the COM Registration Fixing tool.\n"
+		"This tool fixes common problems that sometimes occur, preventing NVDA from accessing content in various "
+		"applications and Windows components.\n"
+		"Installing and uninstalling programs, and other events, can damage accessibility entries in the "
```
I'd move line 488 to 490 before 486-487. I.e. first describe the issue and then explain that the tool will fix these; this more logical to put the cause and then the consequence action. The lines 486-487 will probably need to be rephrased though if they are moved.

Commit:
Update the message again, per review request.

XLTechie commented:
@CyrilleB79 I have updated the message again per your request.
I did have to move some words around and change things a bit more. Hopefully 
this is more satisfactory.

CyrilleB79 commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,90 @@ def onInstallCommand(self, evt):
 
+	_CRFT_INTRO_MESSAGE: str = _(
+		# Translators: Explain the COM Registration Fixing tool to users before running
+		"Welcome to the COM Registration Fixing tool.\n"
```
```suggestion
		"Welcome to the COM Registration Fixing tool.\n\n"
```

CyrilleB79 commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,90 @@ def onInstallCommand(self, evt):
+		'or can cause "unknown" or "pane" to be spoken or brailled in some applications or Windows components, '
+		"instead of the content you were expecting.\n\n'
+		"This tool attempts to fix these common problems.\n"
+		"The tool accesses the system registry, which requires administrative privileges.\n"
```
```suggestion
		"The tool accesses the system registry, which requires administrative privileges.\n\n"
```

CyrilleB79 reviewed (approved):
The text seems quite nice now. Good job!

I have just added as suggestions two more newlines in order to separate the text in 4 paragraphs, what seems visually more appealing to me rather than bigger text blocks in such dialogs. This leads to 4 text blocks:
* welcome sentence
* Description of the situation in which this tool may be useful
* How the tool will proceed, i.e. fix reg entries
* Last sentence to introduce the action buttons, namely "Continue" here.

CyrilleB79 commented:
And by the way, my review was about the dialog's text, but syntax errors should of course be fixed (unmatching quotes and wrong indentation)

SaschaCowley committed:
Syntax fixes

XLTechie committed:
Apply newline suggestions from code review

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

XLTechie commented:
Sascha got to those before I did--not sure what happened with the indent, I probably had indentation set to auto and manually tabbed. A little disturbed that Ruf didn't pick it up in the autofix though.

CyrilleB79 commented:
> Sascha got to those before I did--not sure what happened with the indent, I probably had indentation set to auto and manually tabbed. A little disturbed that Ruf didn't pick it up in the autofix though.

Probably auto-fix was not possible. Ruff may see that the syntax is bad but in the case of tab, it cannot imagine where you want to indent or oudent; it depends on what you want to achieve in your code.

Commit:
Very minor grammar enhancement.

XLTechie commented:
This is ready, I believe.

@Qchristensen I think the dialog content got simplified since your review. If you checked that, you may want to again.

Commit:
Merge branch 'master' into CRFTRefactor

Commit:
Correction to an unrelated changes listing

XLTechie commented on user_docs/en/userGuide.md:
```diff
@@ -3806,19 +3806,23 @@ Follow the directions in [Creating a portable copy](#CreatingAPortableCopy) for
 
 Installing and uninstalling programs on a computer can, in certain cases, cause COM DLL files to get unregistered.
-As COM Interfaces such as IAccessible depend on correct COM DLL registrations, issues can appear in case the correct registration is missing.
+As accessibility interfaces such as IAccessible depend on correct COM DLL registrations, issues can appear if the correct registrations are missing.
```
```suggestion
Because accessibility interfaces such as IAccessible depend on correct COM DLL registrations, irregular behavior can appear if the correct registrations are missing.
```

XLTechie committed:
Slight grammar clarification in user guide.

XLTechie commented on user_docs/en/userGuide.md:
```diff
@@ -3806,19 +3806,23 @@ Follow the directions in [Creating a portable copy](#CreatingAPortableCopy) for
 
 Installing and uninstalling programs on a computer can, in certain cases, cause COM DLL files to get unregistered.
-As COM Interfaces such as IAccessible depend on correct COM DLL registrations, issues can appear in case the correct registration is missing.
+Because accessibility interfaces such as IAccessible depend on correct COM DLL registrations, irregular behavior can appear if the correct registrations are missing.
```
```suggestion
Because accessibility interfaces such as IAccessible depend on correct COM DLL registrations, abnormal behavior can appear if the correct registrations are missing.
```

XLTechie committed:
Correction to grammar clarification in user guide.

SaschaCowley commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,90 @@ def onInstallCommand(self, evt):
+					_("COM Registration Fixing Tool Failed"),
+					wx.OK,
+				)
+		# Display success dialog if there were no errors
```
What stops this from showing if there was an error?

XLTechie replied:
The `return`s in the two `except` blocks above it. It is never executed if there was an error, because the error is handled by the `except`s and shown by the `finally`.

SaschaCowley replied:
Aah, of course 🤦

Commit:
Remove attribution from changes entries

Commit:
Merge branch 'master' into CRFTRefactor

XLTechie commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,90 @@ def onInstallCommand(self, evt):
+		"Windows registry. This can cause previously accessible elements to be presented incorrectly, "
+		'or can cause "unknown" or "pane" to be spoken or brailled in some applications or Windows components, '
+		"instead of the content you were expecting.\n\n"
+		"This tool attempts to fix these common problems.\n"
```
```suggestion
		"This tool attempts to fix these common problems. "
```

XLTechie commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,90 @@ def onInstallCommand(self, evt):
+		'or can cause "unknown" or "pane" to be spoken or brailled in some applications or Windows components, '
+		"instead of the content you were expecting.\n\n"
+		"This tool attempts to fix these common problems.\n"
+		"The tool accesses the system registry, which requires administrative privileges.\n\n"
```
```suggestion
		"Note that the tool must access the system registry, which requires administrative privileges.\n\n"
```

XLTechie committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

Commit:
Update user guide per review suggestion.

Commit:
Add a clarifying comment

XLTechie commented:
@SaschaCowley I have addressed your review. I'm not sure I like the user guide reordered as you requested, but I do agree with your reason for wanting to do it.

XLTechie commented on source/gui/__init__.py:
```diff
@@ -479,50 +480,90 @@ def onInstallCommand(self, evt):
+		"Windows registry. This can cause previously accessible elements to be presented incorrectly, "
+		'or can cause "unknown" or "pane" to be spoken or brailled in some applications or Windows components, '
+		"instead of the content you were expecting.\n\n"
+		"This tool attempts to fix these common problems. "
```
```suggestion
		"This tool attempts to fix such common problems. "
```

XLTechie committed:
Change a word in the intro message for more clarity.

Commit:
Merge branch 'master' into CRFTRefactor

SaschaCowley commented:
@XLTechie I agree it really doesn't read very well now. Perhaps we could rewrite it to be a little less technical and more to the point of what users are likely to care about? What do you think of something like this?

```md
Installing and uninstalling programs on a computer can sometimes cause problems with the Windows registry that result in NVDA behaving abnormally.
This can happen, for example, after installing and uninstalling Adobe Reader, Math Player and other programs.
It can also happen after Windows updates, or during other events that access the registry.
THE COM Registration Fixing Tool attempts to fix these issues by repairing accessibility entries in the registry.

The types of issues this tool can fix include:

* NVDA reporting "unknown" or "pane" when navigating in browsers such as Firefox or Edge, mail programs such as Thunderbird, Windows Explorer, the task bar, and other programs.
* NVDA failing to switch between focus mode and browse mode when you expect it to.
* NVDA being very slow when navigating in browsers while using browse mode.

Because this tool corrects entries in the Windows registry, it requires administrative access to work, just like when installing a program.
If you have UAC (User Access Control) enabled, as most users do, you will need to follow whatever prompts are presented by UAC to run the tool successfully.
```

Commit:
Revert "Update user guide per review suggestion.": a newer version may be used instead.

This reverts commit 78a13985eef59f1d8b74fb46fedae5ff3366f3f0.

Commit:
Merge branch 'master' into CRFTRefactor

Commit:
Potential new user guide entry.

XLTechie commented:
@SaschaCowley How about this? Not quite what you wrote, but closer. @Qchristensen you might want to get in on this.

Commit:
Fix a tiny bit of structure in the UG

Qchristensen reviewed (approved):
Updated user guide and dialog intro text reads well.  Great work everyone
